Install dependencies with Conan

tkoskela opened the pull request:
close #262 

Replaces packages that were installed with `lookup_package` from GreatCmakeCookOff with `conan install`

tkoskela committed:
add conanfile

tkoskela committed:
Add fftw because we need it with some build options

tkoskela committed:
Comment out package lookups because these should be now provided by conan

tkoskela committed:
Add conan commands from step 6 of

https://docs.conan.io/en/latest/getting_started.html#an-md5-hash-calculator-using-the-poco-libraries

tkoskela committed:
Include exception.h from sopt

tkoskela committed:
include find_package, commet out lookup_package for simplicity

SJaffa committed:
Include vector and time libraries that GreatCMakeCookoff was adding somehow

tkoskela committed:
Merge branch 'development' into tk/build

tkoskela committed:
Install conan to test new build system. Stop GA running on draft PRs.

SJaffa committed:
Fixing spdlog linking

tkoskela commented on cmake_files/dependencies.cmake:
```diff
@@ -1,9 +1,9 @@
-if(logging)
-  lookup_package(spdlog REQUIRED)
-endif()
+# lookup_package(Eigen3 REQUIRED ARGUMENTS MD5 9e30f67e8531477de4117506fe44669b URL https://gitlab.com/libeigen/eigen/-/archive/3.3.7/eigen-3.3.7.tar.gz)
+# if(logging)
+#   lookup_package(spdlog REQUIRED)
+# endif()
```
Remove commented lines
```suggestion
```

tkoskela commented on cmake_files/dependencies.cmake:
```diff
@@ -24,7 +24,7 @@ endif()
 
 if(regressions)
-  find_package(FFTW3 REQUIRED DOUBLE)
+  #find_package(FFTW3 REQUIRED DOUBLE)
```
Remove commented lines
```suggestion
```

tkoskela committed:
Install conan with pip on ubuntu. Add conan install step

tkoskela committed:
Install conan for docs for compatibility

tkoskela committed:
Set paths for conan correctly

tkoskela committed:
Set libcxx in conan install to avoid spdlog issue

tkoskela committed:
Build missing packages with conan

tkoskela committed:
Try explicitly running on PRs that are not draft

tkoskela committed:
Separate conan install into gcc and clang. Don't set libcxx on clang

tkoskela committed:
Update cmake_files/dependencies.cmake

Remove comments

tkoskela committed:
Update cmake_files/dependencies.cmake

Remove comments

tkoskela commented:
Building missing packages with conan is very slow. We may want to add them to our own conan repository.

SJaffa commented on cmake_files/dependencies.cmake:
```diff
@@ -24,7 +20,6 @@ endif()
 
```
We are installing tiff through conan, should we remove this bit above?

dpshelio replied:
We may need a `conanfile.py` to [set conditional requirements](https://docs.conan.io/en/latest/mastering/conditional.html) as for tiff and doxygen to happen only when `docs`, `examples` or `regresion` are built.

SJaffa replied:
Done

dpshelio commented on .github/workflows/documentation.yml:
```diff
@@ -15,12 +16,15 @@ jobs:
         
       - name: Configure and build docs
-        run: | 
+        run: |          
```
```suggestion
        run: |
```

dpshelio commented on cmake_files/dependencies.cmake:
```diff
@@ -24,7 +20,6 @@ endif()
 
 
 if(regressions)
-  find_package(FFTW3 REQUIRED DOUBLE)
```
is `fftw3` only used for the regressions? I wouldn't set it as a global dependency.

tkoskela replied:
@SJaffa and I liked the `conanfile.py` idea and added one in [e926cab](https://github.com/astro-informatics/sopt/pull/265/commits/e926cabc0708bf49ea6668f36ee09a4da7919c3d)

dpshelio reviewed (commented):
We may go ahead and merge this PR, but we need to revisit the dependencies and see whether we can set the dependencies, so they are installed only when needed.

Are we sure all these new `#include ctime`s are needed?

tkoskela committed:
Remove whitespace

Co-authored-by: David Pérez-Suárez <dps.helio@gmail.com>

SJaffa committed:
Changed conanfile.txt into python version

SJaffa committed:
Merge branch 'tk/build' of github.com:astro-informatics/sopt into tk/build

tkoskela commented:
My macbook reproduces the macos errors. Half an hour of googling has not improved my understanding of what's going on

```
ctest . --rerun-failed --output-on-failure
Test project /Users/tkoskela/git_repos/sopt/build
      Start  1: test_wavelets
 1/36 Test  #1: test_wavelets ........................................Subprocess aborted***Exception:   0.05 sec
dyld[19259]: missing symbol called

      Start  2: test_sara
 2/36 Test  #2: test_sara ............................................Subprocess aborted***Exception:   0.06 sec
dyld[19260]: missing symbol called

      Start  3: test_maths
 3/36 Test  #3: test_maths ...........................................Subprocess aborted***Exception:   0.06 sec
dyld[19261]: missing symbol called

      Start  4: test_wrapper
 4/36 Test  #4: test_wrapper .........................................Subprocess aborted***Exception:   0.06 sec
dyld[19262]: missing symbol called

      Start  5: test_conjugate_gradient
 5/36 Test  #5: test_conjugate_gradient ..............................Subprocess aborted***Exception:   0.06 sec
dyld[19263]: missing symbol called

      Start  6: test_linear_transform
 6/36 Test  #6: test_linear_transform ................................Subprocess aborted***Exception:   0.06 sec
dyld[19264]: missing symbol called

      Start  7: test_chained_operators
 7/36 Test  #7: test_chained_operators ...............................Subprocess aborted***Exception:   0.06 sec
dyld[19265]: missing symbol called

      Start  8: test_sdmm
 8/36 Test  #8: test_sdmm ............................................Subprocess aborted***Exception:   0.06 sec
dyld[19266]: missing symbol called

      Start  9: test_sdmm_warm_start
 9/36 Test  #9: test_sdmm_warm_start .................................Subprocess aborted***Exception:   0.07 sec
dyld[19267]: missing symbol called

      Start 10: test_proximal
10/36 Test #10: test_proximal ........................................Subprocess aborted***Exception:   0.06 sec
dyld[19268]: missing symbol called

      Start 11: seeded_proximal
11/36 Test #11: seeded_proximal ......................................Subprocess aborted***Exception:   0.06 sec
dyld[19269]: missing symbol called

      Start 12: test_padmm
12/36 Test #12: test_padmm ...........................................Subprocess aborted***Exception:   0.06 sec
dyld[19270]: missing symbol called

      Start 13: test_forward_backward
13/36 Test #13: test_forward_backward ................................Subprocess aborted***Exception:   0.06 sec
dyld[19271]: missing symbol called

      Start 14: test_padmm_warm_start
14/36 Test #14: test_padmm_warm_start ................................Subprocess aborted***Exception:   0.06 sec
dyld[19272]: missing symbol called

      Start 15: test_reweighted
15/36 Test #15: test_reweighted ......................................Subprocess aborted***Exception:   0.06 sec
dyld[19273]: missing symbol called

      Start 16: test_power_method
16/36 Test #16: test_power_method ....................................Subprocess aborted***Exception:   0.06 sec
dyld[19274]: missing symbol called

      Start 17: test_primal_dual
17/36 Test #17: test_primal_dual .....................................Subprocess aborted***Exception:   0.06 sec
dyld[19275]: missing symbol called

      Start 18: test_credible_region
18/36 Test #18: test_credible_region .................................Subprocess aborted***Exception:   0.06 sec
dyld[19276]: missing symbol called

      Start 19: test_bisection_method
19/36 Test #19: test_bisection_method ................................Subprocess aborted***Exception:   0.06 sec
dyld[19277]: missing symbol called

      Start 20: communicator
20/36 Test #20: communicator .........................................***Failed    2.39 sec
dyld[19279]: missing symbol called
[Tuomass-MBP:19279] *** Process received signal ***
[Tuomass-MBP:19279] Signal: Abort trap: 6 (6)
[Tuomass-MBP:19279] Signal code:  (0)
[Tuomass-MBP:19279] [ 0] dyld[19280]: missing symbol called
[Tuomass-MBP:19280] *** Process received signal ***
[Tuomass-MBP:19280] Signal: Abort trap: 6 (6)
[Tuomass-MBP:19280] Signal code:  (0)
[Tuomass-MBP:19280] [ 0] 0   libsystem_platform.dylib            0x00007ff81c5c6e2d _sigtramp + 29
[Tuomass-MBP:19280] [ 1] 0   libsystem_platform.dylib            0x00007ff81c5c6e2d _sigtramp + 29
[Tuomass-MBP:19279] [ 1] 0   ???                                 0x0000000000000000 0x0 + 0
[Tuomass-MBP:19279] [ 2] 0   ???                                 0x0000000000000000 0x0 + 0
[Tuomass-MBP:19280] [ 2] 0   dyld                                0x000000010f911201 dyld + 377345
[Tuomass-MBP:19280] [ 3] 0   dyld                                0x000000010ddee201 dyld + 377345
[Tuomass-MBP:19279] [ 3] 0   dyld                                0x000000010f8bf613 _ZN5dyld44haltEPKc + 375
[Tuomass-MBP:19280] [ 4] 0   dyld                                0x000000010dd9c613 _ZN5dyld44haltEPKc + 375
[Tuomass-MBP:19279] [ 4] 0   dyld                                0x000000010f8dbc6e _ZN5dyld44APIs11_tlv_atexitEPFvPvES1_ + 0
[Tuomass-MBP:19280] [ 5] 0   dyld                                0x000000010ddb8c6e _ZN5dyld44APIs11_tlv_atexitEPFvPvES1_ + 0
[Tuomass-MBP:19279] [ 5] 0   test_communicator                   0x000000010dfa1bb7 _ZN6spdlog5sinks16stdout_sink_baseINS_7details13console_mutexEEC2EP7__sFILE + 103
[Tuomass-MBP:19280] [ 6] 0   test_communicator                   0x000000010630cbb7 _ZN6spdlog5sinks16stdout_sink_baseINS_7details13console_mutexEEC2EP7__sFILE + 103
[Tuomass-MBP:19279] [ 6] 0   test_communicator                   0x000000010dfa24e3 _ZN6spdlog19synchronous_factory6createINS_5sinks11stdout_sinkINS_7details13console_mutexEEEJEEENSt3__110shared_ptrINS_6loggerEEENS7_12basic_stringIcNS7_11char_traitsIcEENS7_9allocatorIcEEEEDpOT0_ + 83
[Tuomass-MBP:19280] [ 7] 0   test_communicator                   0x000000010dfa2455 _ZN6spdlog16stdout_logger_mtINS_19synchronous_factoryEEENSt3__110shared_ptrINS_6loggerEEERKNS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEE + 37
[Tuomass-MBP:19280] [ 8] 0   test_communicator                   0x000000010630d4e3 _ZN6spdlog19synchronous_factory6createINS_5sinks11stdout_sinkINS_7details13console_mutexEEEJEEENSt3__110shared_ptrINS_6loggerEEENS7_12basic_stringIcNS7_11char_traitsIcEENS7_9allocatorIcEEEEDpOT0_ + 83
[Tuomass-MBP:19279] [ 7] 0   test_communicator                   0x000000010630d455 _ZN6spdlog16stdout_logger_mtINS_19synchronous_factoryEEENSt3__110shared_ptrINS_6loggerEEERKNS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEE + 37
[Tuomass-MBP:19279] [ 8] 0   test_communicator                   0x000000010df7345d _ZN4sopt7logging10initializeERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 109
[Tuomass-MBP:19280] [ 9] 0   test_communicator                   0x000000010df730a3 main + 1507
[Tuomass-MBP:19280] [10] 0   test_communicator                   0x00000001062de45d _ZN4sopt7logging10initializeERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 109
[Tuomass-MBP:19279] [ 9] 0   test_communicator                   0x00000001062de0a3 main + 1507
[Tuomass-MBP:19279] [10] 0   dyld                                0x000000010f8ba4fe start + 462
[Tuomass-MBP:19280] *** End of error message ***
0   dyld                                0x000000010dd974fe start + 462
[Tuomass-MBP:19279] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 1 with PID 0 on node Tuomass-MBP exited on signal 6 (Abort trap: 6).
--------------------------------------------------------------------------

      Start 21: mpi_proximals
21/36 Test #21: mpi_proximals ........................................***Failed    2.36 sec
dyld[19282]: missing symbol called
[Tuomass-MBP:19282] *** Process received signal ***
[Tuomass-MBP:19282] Signal: Abort trap: 6 (6)
[Tuomass-MBP:19282] Signal code:  (0)
[Tuomass-MBP:19282] [ 0] 0   libsystem_platform.dylib            0x00007ff81c5c6e2d _sigtramp + 29
[Tuomass-MBP:19282] [ 1] 0   ???                                 0x0000000000000000 0x0 + 0
[Tuomass-MBP:19282] [ 2] 0   dyld                                0x0000000117825201 dyld + 377345
[Tuomass-MBP:19282] [ 3] dyld[19283]: missing symbol called
[Tuomass-MBP:19283] *** Process received signal ***
[Tuomass-MBP:19283] Signal: Abort trap: 6 (6)
[Tuomass-MBP:19283] Signal code:  (0)
[Tuomass-MBP:19283] [ 0] 0   libsystem_platform.dylib            0x00007ff81c5c6e2d _sigtramp + 29
[Tuomass-MBP:19283] [ 1] 0   ???                                 0x0000000000000000 0x0 + 0
[Tuomass-MBP:19283] [ 2] 0   dyld                                0x00000001177d3613 _ZN5dyld44haltEPKc + 375
[Tuomass-MBP:19282] [ 4] 0   dyld                                0x00000001177efc6e _ZN5dyld44APIs11_tlv_atexitEPFvPvES1_ + 0
[Tuomass-MBP:19282] [ 5] 0   dyld                                0x000000011148e201 dyld + 377345
[Tuomass-MBP:19283] [ 3] 0   test_mpi_proximals                  0x000000010d1e3797 _ZN6spdlog5sinks16stdout_sink_baseINS_7details13console_mutexEEC2EP7__sFILE + 103
[Tuomass-MBP:19282] [ 6] 0   test_mpi_proximals                  0x000000010d1e40c3 _ZN6spdlog19synchronous_factory6createINS_5sinks11stdout_sinkINS_7details13console_mutexEEEJEEENSt3__110shared_ptrINS_6loggerEEENS7_12basic_stringIcNS7_11char_traitsIcEENS7_9allocatorIcEEEEDpOT0_ + 83
[Tuomass-MBP:19282] [ 7] 0   dyld                                0x000000011143c613 _ZN5dyld44haltEPKc + 375
[Tuomass-MBP:19283] [ 4] 0   test_mpi_proximals                  0x000000010d1e4035 _ZN6spdlog16stdout_logger_mtINS_19synchronous_factoryEEENSt3__110shared_ptrINS_6loggerEEERKNS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEE + 37
[Tuomass-MBP:19282] [ 8] 0   test_mpi_proximals                  0x000000010d1b44bd _ZN4sopt7logging10initializeERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 109
[Tuomass-MBP:19282] [ 9] 0   test_mpi_proximals                  0x000000010d1b4103 main + 1507
[Tuomass-MBP:19282] [10] 0   dyld                                0x0000000111458c6e _ZN5dyld44APIs11_tlv_atexitEPFvPvES1_ + 0
[Tuomass-MBP:19283] [ 5] 0   test_mpi_proximals                  0x0000000104571797 _ZN6spdlog5sinks16stdout_sink_baseINS_7details13console_mutexEEC2EP7__sFILE + 103
[Tuomass-MBP:19283] [ 6] 0   dyld                                0x00000001177ce4fe start + 462
[Tuomass-MBP:19282] *** End of error message ***
0   test_mpi_proximals                  0x00000001045720c3 _ZN6spdlog19synchronous_factory6createINS_5sinks11stdout_sinkINS_7details13console_mutexEEEJEEENSt3__110shared_ptrINS_6loggerEEENS7_12basic_stringIcNS7_11char_traitsIcEENS7_9allocatorIcEEEEDpOT0_ + 83
[Tuomass-MBP:19283] [ 7] 0   test_mpi_proximals                  0x0000000104572035 _ZN6spdlog16stdout_logger_mtINS_19synchronous_factoryEEENSt3__110shared_ptrINS_6loggerEEERKNS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEE + 37
[Tuomass-MBP:19283] [ 8] 0   test_mpi_proximals                  0x00000001045424bd _ZN4sopt7logging10initializeERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 109
[Tuomass-MBP:19283] [ 9] 0   test_mpi_proximals                  0x0000000104542103 main + 1507
[Tuomass-MBP:19283] [10] 0   dyld                                0x00000001114374fe start + 462
[Tuomass-MBP:19283] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 1 with PID 0 on node Tuomass-MBP exited on signal 6 (Abort trap: 6).
--------------------------------------------------------------------------

      Start 22: serial_vs_parallel_padmm
22/36 Test #22: serial_vs_parallel_padmm .............................***Failed    2.36 sec
dyld[19285]: missing symbol called
dyld[19286]: missing symbol called
[Tuomass-MBP:19286] *** Process received signal ***
[Tuomass-MBP:19286] Signal: Abort trap: 6 (6)
[Tuomass-MBP:19286] Signal code:  (0)
[Tuomass-MBP:19286] [ 0] 0   libsystem_platform.dylib            0x00007ff81c5c6e2d _sigtramp + 29
[Tuomass-MBP:19286] [ 1] 0   ???                                 0x0000000000000000 0x0 + 0
[Tuomass-MBP:19286] [ 2] 0   dyld                                0x000000011a61b201 dyld + 377345
[Tuomass-MBP:19286] [ 3] 0   dyld                                0x000000011a5c9613 _ZN5dyld44haltEPKc + 375
[Tuomass-MBP:19286] [ 4] [Tuomass-MBP:19285] *** Process received signal ***
[Tuomass-MBP:19285] Signal: Abort trap: 6 (6)
[Tuomass-MBP:19285] Signal code:  (0)
[Tuomass-MBP:19285] [ 0] 0   libsystem_platform.dylib            0x00007ff81c5c6e2d _sigtramp + 29
[Tuomass-MBP:19285] [ 1] 0   ???                                 0x0000000000000000 0x0 + 0
[Tuomass-MBP:19285] [ 2] 0   dyld                                0x000000010ae58201 dyld + 377345
[Tuomass-MBP:19285] [ 3] 0   dyld                                0x000000010ae06613 _ZN5dyld44haltEPKc + 375
[Tuomass-MBP:19285] [ 4] 0   dyld                                0x000000011a5e5c6e _ZN5dyld44APIs11_tlv_atexitEPFvPvES1_ + 0
[Tuomass-MBP:19286] [ 5] 0   dyld                                0x000000010ae22c6e _ZN5dyld44APIs11_tlv_atexitEPFvPvES1_ + 0
[Tuomass-MBP:19285] [ 5] 0   test_serial_vs_parallel_padmm       0x000000010f82e407 _ZN6spdlog5sinks16stdout_sink_baseINS_7details13console_mutexEEC2EP7__sFILE + 103
[Tuomass-MBP:19286] [ 6] 0   test_serial_vs_parallel_padmm       0x000000010f82ed33 _ZN6spdlog19synchronous_factory6createINS_5sinks11stdout_sinkINS_7details13console_mutexEEEJEEENSt3__110shared_ptrINS_6loggerEEENS7_12basic_stringIcNS7_11char_traitsIcEENS7_9allocatorIcEEEEDpOT0_ + 83
[Tuomass-MBP:19286] [ 7] 0   test_serial_vs_parallel_padmm       0x000000010086f407 _ZN6spdlog5sinks16stdout_sink_baseINS_7details13console_mutexEEC2EP7__sFILE + 103
[Tuomass-MBP:19285] [ 6] 0   test_serial_vs_parallel_padmm       0x000000010086fd33 _ZN6spdlog19synchronous_factory6createINS_5sinks11stdout_sinkINS_7details13console_mutexEEEJEEENSt3__110shared_ptrINS_6loggerEEENS7_12basic_stringIcNS7_11char_traitsIcEENS7_9allocatorIcEEEEDpOT0_ + 83
[Tuomass-MBP:19285] [ 7] 0   test_serial_vs_parallel_padmm       0x000000010f82eca5 _ZN6spdlog16stdout_logger_mtINS_19synchronous_factoryEEENSt3__110shared_ptrINS_6loggerEEERKNS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEE + 37
[Tuomass-MBP:19286] [ 8] 0   test_serial_vs_parallel_padmm       0x000000010f7ffc1d _ZN4sopt7logging10initializeERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 109
[Tuomass-MBP:19286] [ 9] 0   test_serial_vs_parallel_padmm       0x000000010086fca5 _ZN6spdlog16stdout_logger_mtINS_19synchronous_factoryEEENSt3__110shared_ptrINS_6loggerEEERKNS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEE + 37
[Tuomass-MBP:19285] [ 8] 0   test_serial_vs_parallel_padmm       0x000000010f7ff863 main + 1507
[Tuomass-MBP:19286] [10] 0   test_serial_vs_parallel_padmm       0x0000000100840c1d _ZN4sopt7logging10initializeERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 109
[Tuomass-MBP:19285] [ 9] 0   test_serial_vs_parallel_padmm       0x0000000100840863 main + 1507
[Tuomass-MBP:19285] [10] 0   dyld                                0x000000011a5c44fe start + 462
[Tuomass-MBP:19286] *** End of error message ***
0   dyld                                0x000000010ae014fe start + 462
[Tuomass-MBP:19285] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 1 with PID 0 on node Tuomass-MBP exited on signal 6 (Abort trap: 6).
--------------------------------------------------------------------------

      Start 23: mpi_wavelets
23/36 Test #23: mpi_wavelets .........................................***Failed    2.36 sec
dyld[19293]: missing symbol called
[Tuomass-MBP:19293] *** Process received signal ***
dyld[19292]: missing symbol called
[Tuomass-MBP:19292] *** Process received signal ***
[Tuomass-MBP:19292] Signal: Abort trap: 6 (6)
[Tuomass-MBP:19292] Signal code:  (0)
[Tuomass-MBP:19292] [ 0] 0   libsystem_platform.dylib            0x00007ff81c5c6e2d _sigtramp + 29
[Tuomass-MBP:19292] [ 1] 0   ???                                 0x0000000000000000 0x0 + 0
[Tuomass-MBP:19292] [ 2] [Tuomass-MBP:19293] Signal: Abort trap: 6 (6)
[Tuomass-MBP:19293] Signal code:  (0)
[Tuomass-MBP:19293] [ 0] 0   libsystem_platform.dylib            0x00007ff81c5c6e2d _sigtramp + 29
[Tuomass-MBP:19293] [ 1] 0   ???                                 0x0000000000000000 0x0 + 0
[Tuomass-MBP:19293] [ 2] 0   dyld                                0x0000000112970201 dyld + 377345
0   dyld                                0x000000010e6ad201 dyld + 377345
[Tuomass-MBP:19292] [ 3] [Tuomass-MBP:19293] [ 3] 0   dyld                                0x000000010e65b613 _ZN5dyld44haltEPKc + 375
[Tuomass-MBP:19292] [ 4] 0   dyld                                0x000000011291e613 _ZN5dyld44haltEPKc + 375
[Tuomass-MBP:19293] [ 4] 0   dyld                                0x000000011293ac6e _ZN5dyld44APIs11_tlv_atexitEPFvPvES1_ + 0
[Tuomass-MBP:19293] [ 5] 0   dyld                                0x000000010e677c6e _ZN5dyld44APIs11_tlv_atexitEPFvPvES1_ + 0
[Tuomass-MBP:19292] [ 5] 0   test_mpi_wavelets                   0x00000001089a1a67 _ZN6spdlog5sinks16stdout_sink_baseINS_7details13console_mutexEEC2EP7__sFILE + 103
[Tuomass-MBP:19292] [ 6] 0   test_mpi_wavelets                   0x00000001081e4a67 _ZN6spdlog5sinks16stdout_sink_baseINS_7details13console_mutexEEC2EP7__sFILE + 103
[Tuomass-MBP:19293] [ 6] 0   test_mpi_wavelets                   0x00000001081e5393 _ZN6spdlog19synchronous_factory6createINS_5sinks11stdout_sinkINS_7details13console_mutexEEEJEEENSt3__110shared_ptrINS_6loggerEEENS7_12basic_stringIcNS7_11char_traitsIcEENS7_9allocatorIcEEEEDpOT0_ + 83
[Tuomass-MBP:19293] [ 7] 0   test_mpi_wavelets                   0x00000001081e5305 _ZN6spdlog16stdout_logger_mtINS_19synchronous_factoryEEENSt3__110shared_ptrINS_6loggerEEERKNS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEE + 37
[Tuomass-MBP:19293] [ 8] 0   test_mpi_wavelets                   0x00000001089a2393 _ZN6spdlog19synchronous_factory6createINS_5sinks11stdout_sinkINS_7details13console_mutexEEEJEEENSt3__110shared_ptrINS_6loggerEEENS7_12basic_stringIcNS7_11char_traitsIcEENS7_9allocatorIcEEEEDpOT0_ + 83
[Tuomass-MBP:19292] [ 7] 0   test_mpi_wavelets                   0x00000001089a2305 _ZN6spdlog16stdout_logger_mtINS_19synchronous_factoryEEENSt3__110shared_ptrINS_6loggerEEERKNS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEE + 37
[Tuomass-MBP:19292] [ 8] 0   test_mpi_wavelets                   0x00000001089727bd _ZN4sopt7logging10initializeERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 109
[Tuomass-MBP:19292] [ 9] 0   test_mpi_wavelets                   0x00000001081b57bd _ZN4sopt7logging10initializeERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 109
[Tuomass-MBP:19293] [ 9] 0   test_mpi_wavelets                   0x0000000108972403 main + 1507
[Tuomass-MBP:19292] [10] 0   test_mpi_wavelets                   0x00000001081b5403 main + 1507
[Tuomass-MBP:19293] [10] 0   dyld                                0x000000010e6564fe start + 462
[Tuomass-MBP:19292] *** End of error message ***
0   dyld                                0x00000001129194fe start + 462
[Tuomass-MBP:19293] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 1 with PID 0 on node Tuomass-MBP exited on signal 6 (Abort trap: 6).
--------------------------------------------------------------------------

      Start 24: mpi_session
24/36 Test #24: mpi_session ..........................................***Failed    2.23 sec
dyld[19295]: missing symbol called
dyld[19296]: missing symbol called
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 1 with PID 0 on node Tuomass-MBP exited on signal 6 (Abort trap: 6).
--------------------------------------------------------------------------

      Start 25: test_example_wavelets
25/36 Test #25: test_example_wavelets ................................Subprocess aborted***Exception:   0.04 sec
dyld[19297]: missing symbol called

      Start 26: test_example_sara
26/36 Test #26: test_example_sara ....................................Subprocess aborted***Exception:   0.05 sec
dyld[19298]: missing symbol called

      Start 30: test_example_conjugate_gradient
27/36 Test #30: test_example_conjugate_gradient ......................Subprocess aborted***Exception:   0.06 sec
dyld[19299]: missing symbol called

      Start 31: test_example_l1_proximal
28/36 Test #31: test_example_l1_proximal .............................Subprocess aborted***Exception:   0.05 sec
dyld[19300]: missing symbol called

      Start 32: test_example_power_method
29/36 Test #32: test_example_power_method ............................Subprocess aborted***Exception:   0.06 sec
dyld[19301]: missing symbol called

      Start 33: test_example_euclidian_norm
30/36 Test #33: test_example_euclidian_norm ..........................Subprocess aborted***Exception:   0.05 sec
dyld[19302]: missing symbol called

      Start 34: test_example_inpainting
31/36 Test #34: test_example_inpainting ..............................Subprocess aborted***Exception:   0.06 sec
dyld[19303]: missing symbol called

      Start 35: test_example_padmm_euclidian_norm
32/36 Test #35: test_example_padmm_euclidian_norm ....................Subprocess aborted***Exception:   0.06 sec
dyld[19304]: missing symbol called

      Start 36: test_example_padmm_inpainting
33/36 Test #36: test_example_padmm_inpainting ........................Subprocess aborted***Exception:   0.06 sec
dyld[19305]: missing symbol called

      Start 37: test_example_pd_inpainting
34/36 Test #37: test_example_pd_inpainting ...........................Subprocess aborted***Exception:   0.05 sec
dyld[19306]: missing symbol called

      Start 38: test_example_forward_backward_inpainting
35/36 Test #38: test_example_forward_backward_inpainting .............Subprocess aborted***Exception:   0.06 sec
dyld[19307]: missing symbol called

      Start 39: test_example_forward_backward_inpainting_joint_map
36/36 Test #39: test_example_forward_backward_inpainting_joint_map ...Subprocess aborted***Exception:   0.05 sec
dyld[19308]: missing symbol called


0% tests passed, 36 tests failed out of 36
```

tkoskela committed:
Build with conan in github actions workflow.
Move ccache cmake flags to conanfile.py.

tkoskela committed:
Use -bf instead of -if in build. Pass mpi and omp to conan install.

tkoskela committed:
Fix clang build by adding missing -o

tkoskela committed:
Use KEEP_RPATHS to fix broken dynamic links on MacOS, see

https://docs.conan.io/en/latest/howtos/manage_shared_libraries/rpaths.html

tkoskela commented:
On [41dae34](https://github.com/astro-informatics/sopt/pull/265/commits/41dae3406fe8b6d47f16e801b2293ecefdecb57e) test #8 failed on my Mac on the first time I ran it. All following runs passed so I could not reproduce. Could this be caused by the random seed? @MatthijsMars @CosmoMatt 

```ctest
8/39 Test  #8: test_sdmm ............................................***Failed    0.49 sec
[2022-03-18 11:12:38.174] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.203] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.244] [sopt] [critical] Performing SDMM 

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_sdmm is a Catch v2.13.7 host application.
Run with -? for options

Randomness seeded to: 1647601958

-------------------------------------------------------------------------------
SDMM with ||x - x0||_2 functions
  With different L
-------------------------------------------------------------------------------
/Users/tkoskela/git_repos/sopt/cpp/tests/sdmm.cc:211
...............................................................................

/Users/tkoskela/git_repos/sopt/cpp/tests/sdmm.cc:229: FAILED:
  CHECK( func(result) <= func(result + epsilon) )
with expansion:
  3.7534305808 <= 3.7534304842
with messages:
  result.transpose() :=    -1.647  0.311289   0.87681 -0.863347
  epsilon.transpose() :=      0 0.0001      0      0

/Users/tkoskela/git_repos/sopt/cpp/tests/sdmm.cc:230: FAILED:
  CHECK( func(result) <= func(result - epsilon) )
with expansion:
  3.7534305808 <= 3.7534304071
with messages:
  result.transpose() :=    -1.647  0.311289   0.87681 -0.863347
  epsilon.transpose() :=      0      0      0 0.0001

===============================================================================
test cases:  3 |  2 passed | 1 failed
assertions: 85 | 83 passed | 2 failed
```

MatthijsMars commented:
@tkoskela Looking at the test I think that might possibly be caused by a bad random seed. Still, I think it is kind of strange since, this should always converge to a solution for a simple problem like this, but if the fault cannot be reproduced it might have been a particularly bad operator or data vector.

SJaffa commented on CMakeLists.txt:
```diff
@@ -42,8 +42,12 @@ cxx11_feature_check(REQUIRED unique_ptr nullptr override constructor_delegate)
 include(compilations)
 
 # search/install dependencies
+# with lookup_package calls commented out
```
```suggestion
```

SJaffa commented on CMakeLists.txt:
```diff
@@ -66,7 +70,8 @@ if(tests OR examples)
 
 if(tests)
-  find_package(Catch 2.3.0 EXACT)
+  #find_package(Catch 2.3.0 EXACT)
```
```suggestion
```

SJaffa committed:
Apply suggestions from code review

Remove commented out sections

SJaffa committed:
Updated installation instructions for Conan install and build

tkoskela commented on README.md:
```diff
@@ -62,18 +58,22 @@ Then, the program can be built with standard `CMake` command:
 cd build
-cmake -DCMAKE_BUILD_TYPE=Release ..
-make
+conan install --build-missing
```
```suggestion
conan install .. --build missing
```

tkoskela commented on README.md:
```diff
@@ -62,18 +58,22 @@ Then, the program can be built with standard `CMake` command:
-cmake -DCMAKE_INSTALL_PREFIX=/X ..
-make install
+conan install --build-missing -if BUILD_FOLDER
+conan build -bf BUILD_FOLDER
```
```suggestion
conan build .. -bf BUILD_FOLDER -if .
```

tkoskela committed:
Define CMAKE_VERBOSE_MAKEFILE (off by default)

tkoskela committed:
Merge branch 'tk/build' of github.com:astro-informatics/sopt into tk/build

tkoskela committed:
Fix install commands

SJaffa commented on README.md:
```diff
@@ -62,18 +58,22 @@ Then, the program can be built with standard `CMake` command:
-cmake -DCMAKE_BUILD_TYPE=Release ..
-make
+conan install .. --build missing
+conan build
```
```suggestion
conan build ..
```

tkoskela committed:
Remove duplicate install instruction

tkoskela committed:
Reduce number of threads to make MPI+OpenMP tests faster.

Remove unnecessary --rerun-failed since we won't know which tests
failed on previous runs

tkoskela commented on cmake_files/dependencies.cmake:
```diff
@@ -1,9 +1,5 @@
 include(PackageLookup)  # check for existence, or install external projects
 
-lookup_package(Eigen3 REQUIRED ARGUMENTS MD5 9e30f67e8531477de4117506fe44669b URL https://gitlab.com/libeigen/eigen/-/archive/3.3.7/eigen-3.3.7.tar.gz)
-if(logging)
-  lookup_package(spdlog REQUIRED)
-endif()
```
The package installation is being handled by conan but I would think we'd still need to call `find_package()` for these + FFTW3 below.

The tests build and pass currently with these lines removed, which is a bit strange.

TODO: Build with regressions=on and see if we get an error about FFTW3
TODO: Build with logging=on
TODO: Add `find_package(Eigen3)` and `find_package(spdlog)`

tkoskela replied:
Switching `CMAKE_VERBOSE_MAKEFILE=ON` again, I can see that conan inserts the paths to all the dependencies it has installed to the compiler flags directly. A compile command looks something like

```
[ 76%] Building CXX object cpp/examples/CMakeFiles/example_l1_proximal.dir/l1_proximal.cc.o
cd /Users/tkoskela/git_repos/sopt/build/cpp/examples && ccache /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++  -I/Users/tkoskela/.conan/data/eigen/3.3.7/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include/eigen3 -I/Users/tkoskela/.conan/data/catch2/2.13.7/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include -I/Users/tkoskela/.conan/data/benchmark/1.6.0/_/_/package/9e63fd2936c0b4a43222b423ad03ee6f6975cbfb/include -I/Users/tkoskela/.conan/data/libtiff/4.0.9/_/_/package/9db7415868fa23bc26464403e5de51da02c50878/include -I/Users/tkoskela/.conan/data/spdlog/1.9.2/_/_/package/2c8ec9b22128f36867b60dd65efb316eb93c9155/include -I/Users/tkoskela/.conan/data/zlib/1.2.11/_/_/package/e4fbacdee83eb1a2e68227ffe19f6e9a61b36cbc/include -I/Users/tkoskela/.conan/data/xz_utils/5.2.5/_/_/package/e4fbacdee83eb1a2e68227ffe19f6e9a61b36cbc/include -I/Users/tkoskela/.conan/data/libjpeg/9d/_/_/package/e4fbacdee83eb1a2e68227ffe19f6e9a61b36cbc/include -I/Users/tkoskela/.conan/data/jbig/20160605/_/_/package/004c156104209a65ece6cc9d71d8115366dd9d08/include -I/Users/tkoskela/.conan/data/fmt/8.0.1/_/_/package/db14d83fb6d5b354ad734851c8c78016652c0ce0/include -I/Users/tkoskela/git_repos/sopt/cpp/sopt/.. -I/Users/tkoskela/git_repos/sopt/build/include -isystem /usr/local/include -isystem /usr/local/Cellar/open-mpi/4.1.2/include -std=c++11   -O3 -DNDEBUG  -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -mmacosx-version-min=12.2 -DLIBJPEG_STATIC -DLZMA_API_STATIC -DSPDLOG_COMPILED_LIB -DSPDLOG_FMT_EXTERNAL "-Xclang -fopenmp" -MD -MT cpp/examples/CMakeFiles/example_l1_proximal.dir/l1_proximal.cc.o -MF CMakeFiles/example_l1_proximal.dir/l1_proximal.cc.o.d -o CMakeFiles/example_l1_proximal.dir/l1_proximal.cc.o -c /Users/tkoskela/git_repos/sopt/cpp/examples/l1_proximal.cc
```

So do I understand correctly then that we do not need to call `find_package()` in cmake at all? @dpshelio

tkoskela replied:
We could try [cmake_find_package_generator](https://docs.conan.io/en/latest/integrations/build_system/cmake/cmake_find_package_generator.html)

SJaffa replied:
Built on Ubuntu with logging=on builds fine, tests 1-21 pass, test 22 took too long so I killed it but will try again later
